bitcoin/src/bip32.rs

```diff
@@ -125,6 +125,39 @@ pub enum ChildNumber {
         index: u32,
     },
 }
+
+/// NormalChildNumber is a ChildNumber that is not hardened
+/// made only to be used internally in `derive_pub()` and `NormalChildIterator`
+#[derive(Copy, Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
+pub struct NormalChildNumber(pub u32);
+
+/// A iterator over the children of a DerivationPath trying to convert them into NormalChildNumber
+/// made only to be used internally in `derive_pub()`
+pub struct NormalChildIterator {
+    base: Vec<NormalChildNumber>,
+}
+impl Into<NormalChildIterator> for DerivationPath {
+    fn into(self) -> NormalChildIterator {
+        NormalChildIterator {
+            base: self
+                .0
+                .into_iter()
+                .filter_map(|child| {
+                    if let ChildNumber::Normal { index } = child {
+                        Some(NormalChildNumber(index))
+                    } else {
+                        None
+                    }
+                })
+                .collect(),
+        }
+    }
+}
+impl Iterator for NormalChildIterator {
+    type Item = NormalChildNumber;
+    fn next(&mut self) -> Option<Self::Item> { self.base.pop() }
```

Comment:
This is iterating backwards.

Reply:
Indeed. And the test battery passed.

Maybe some changes on it ?